src/persist/src/indexed/future.rs

```diff
@@ -0,0 +1,170 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! A persistent, compacting data structure of `(Key, Value, Time, Diff)`
+//! updates, indexed by time.
+
+use std::sync::Arc;
+
+use timely::progress::Antichain;
+use timely::progress::Timestamp;
+
+use crate::error::Error;
+use crate::indexed::cache::BlobCache;
+use crate::indexed::BlobFutureBatch;
+use crate::indexed::BlobFutureMeta;
+use crate::persister::Snapshot;
+use crate::storage::Blob;
+use crate::storage::SeqNo;
+
+/// A persistent, compacting data structure containing `(Key, Value, Time,
+/// Diff)` entries indexed by `(time, key, value)`.
+///
+/// Invariants:
+/// - All entries are after some time frontier and less than some SeqNo.
+///   Intuitively, this is the data that has been drained out of a
+///   [crate::storage::Buffer] but not yet "seal"ed into a
+///   [crate::indexed::trace::BlobTrace].
+/// - TODO: Space usage.
+pub struct BlobFuture {
+    frontier: Antichain<u64>,
+    // NB: The SeqNo ranges here are sorted and contiguous.
```

Comment:
it would be great to know if the ranges are meant to be `[lower, upper)` or `[lower, upper]` or something else

Reply:
This has been switched to Description at your suggestion, which precisely defines these as `[lower, upper)`. I've also thrown a couple comments denormalizing that info here and some other places it seemed relevant.